### ISSUE TITLE
Async support not compatible with alternative event loop implementations like `uvloop`

### ISSUE THREAD

**Issue by citreae535** [oracle/python-oracledb#276]
1. What versions are you using?
`python-oracledb 2.0.0` on Python 3.12 and `Oracle Database XE 18c`.  Both run inside docker containers on Windows 11 23H2 with WSL 2 backend. The database container image is [gvenzl/oracle-xe:18-faststart](https://hub.docker.com/r/gvenzl/oracle-xe).

2. Is it an error or a hang or a crash?
Error.

3. What error(s) or behavior you are seeing?
Using `connect_async()` with [uvloop](https://uvloop.readthedocs.io/user/index.html) (a faster drop-in replacement for `asyncio`) throws the following exception:
(The first line is the success output of `asyncio.run(main())`.)
```
('TEST',) 
Traceback (most recent call last):
  File "src/oracledb/impl/thin/connection.pyx", line 502, in oracledb.thin_impl.AsyncThinConnImpl._connect_with_address
  File "src/oracledb/impl/thin/protocol.pyx", line 567, in _connect_phase_one
  File "src/oracledb/impl/thin/protocol.pyx", line 730, in _connect_tcp
  File "src/oracledb/impl/thin/transport.pyx", line 301, in oracledb.thin_impl.Transport.set_from_socket
AttributeError: 'uvloop.loop.TCPTransport' object has no attribute 'setsockopt'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<frozen runpy>", line 189, in _run_module_as_main
  File "<frozen runpy>", line 148, in _get_module_details
  File "<frozen runpy>", line 112, in _get_module_details
  File "/app/example_service/__init__.py", line 18, in <module>
    uvloop.run(main())
  File "/app/.venv/lib/python3.12/site-packages/uvloop/__init__.py", line 109, in run
    return __asyncio.run(
           ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "uvloop/loop.pyx", line 1517, in uvloop.loop.Loop.run_until_complete
  File "/app/.venv/lib/python3.12/site-packages/uvloop/__init__.py", line 61, in wrapper
    return await main
           ^^^^^^^^^^
  File "/app/example_service/__init__.py", line 8, in main
    async with oracledb.connect_async(
  File "/app/.venv/lib/python3.12/site-packages/oracledb/connection.py", line 1398, in __aenter__
    await self._connect_coroutine
  File "/app/.venv/lib/python3.12/site-packages/oracledb/connection.py", line 1443, in _connect
    await impl.connect(params_impl)
  File "src/oracledb/impl/thin/connection.pyx", line 609, in connect
  File "src/oracledb/impl/thin/connection.pyx", line 605, in oracledb.thin_impl.AsyncThinConnImpl.connect
  File "src/oracledb/impl/thin/connection.pyx", line 563, in _connect_with_params
  File "src/oracledb/impl/thin/connection.pyx", line 543, in _connect_with_description
  File "src/oracledb/impl/thin/connection.pyx", line 512, in _connect_with_address
  File "/app/.venv/lib/python3.12/site-packages/oracledb/errors.py", line 162, in _raise_err
    raise exc_type(_Error(message)) from cause
oracledb.exceptions.OperationalError: DPY-6005: cannot connect to database (CONNECTION_ID=+o4qjIZQoVmu5EZP54G9xw==).
'uvloop.loop.TCPTransport' object has no attribute 'setsockopt'
```

4. Does your application call init_oracle_client()?
No.

5. Include a runnable Python script that shows the problem.
```python
import asyncio

import oracledb
import uvloop


async def main():
    async with oracledb.connect_async(
        user="test", password="test", dsn="host.docker.internal/xepdb1"
    ) as connection:
        with connection.cursor() as cursor:
            await cursor.execute("select user from dual")
            async for result in cursor:
                print(result)


asyncio.run(main())
uvloop.run(main())

```

6. Cause
The thin mode transport implementation detects Async mode by checking if the supplied `transport` is an instance of `asyncio.Transport`. To support alternative implementations without depending on them directly, we need to detect Async mode without hard-coded runtime instance checks. Maybe passing in an explicit flag variable?
https://github.com/oracle/python-oracledb/blob/9645d4f9b3fb196925ac364e6dec3ab953fca222/src/oracledb/impl/thin/transport.pyx#L287-L301

**Comment by anthony-tuininga** [comment]
I have pushed a patch that should add this support. If you are able to [build from source](https://python-oracledb.readthedocs.io/en/latest/user_guide/installation.html#installing-from-source-code) you can verify that it works for you.

**Comment by citreae535** [comment]
I've confirmed that the patch works. Thanks a lot. Can we expect a patch release soon?

**Comment by anthony-tuininga** [comment]
Definitely! I am hoping to make a patch release sometime in the next week or so.

**Comment by anthony-tuininga** [comment]
The patch has been included in version 2.0.1 which was just released.
